k2/csrc/fsa_algo.cc

```diff
@@ -540,4 +542,85 @@ bool TopSort(const Fsa &a, Fsa *b,
   return true;
 }
 
+std::unique_ptr<Fsa> CreateFsa(const std::vector<Arc> &arcs) {
+  if (arcs.empty()) return nullptr;
+
+  std::vector<std::vector<Arc>> vec;
+  for (const auto &arc : arcs) {
+    auto src_state = arc.src_state;
+    auto dest_state = arc.dest_state;
+    auto new_size = std::max(src_state, dest_state);
+    if (new_size >= vec.size()) vec.resize(new_size + 1);
+    vec[src_state].push_back(arc);
+  }
+
+  std::stack<DfsState> stack;
+  std::vector<char> state_status(vec.size(), kNotVisited);
+  std::vector<int32_t> order;
+
+  auto num_states = static_cast<int32_t>(vec.size());
+  for (auto i = 0; i != num_states; ++i) {
+    if (state_status[i] == kVisited) continue;
+    stack.push({i, 0, static_cast<int32_t>(vec[i].size())});
+    state_status[i] = kVisiting;
+    while (!stack.empty()) {
+      auto &current_state = stack.top();
+      auto state = current_state.state;
+
+      if (current_state.arc_begin == current_state.arc_end) {
+        state_status[state] = kVisited;
+        order.push_back(state);
+        stack.pop();
+        continue;
+      }
+
+      const auto &arc = vec[state][current_state.arc_begin];
+      auto next_state = arc.dest_state;
+      auto status = state_status[next_state];
+      switch (status) {
+        case kNotVisited:
+          state_status[next_state] = kVisiting;
+          stack.push(
+              {next_state, 0, static_cast<int32_t>(vec[next_state].size())});
+          ++current_state.arc_begin;
+          break;
+        case kVisiting:
+          LOG(FATAL) << "there is a cycle: " << state << " -> " << next_state;
+          break;
+        case kVisited:
+          ++current_state.arc_begin;
+          break;
+        default:
+          LOG(FATAL) << "Unreachable code is executed!";
+          break;
+      }
+    }
+  }
+
+  CHECK_EQ(num_states, static_cast<int32_t>(order.size()));
```

Comment:
if you keep `num_states` with `size_t`, it will be not needed to convert type here again.